File: CS/HierarchyToDataSet.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using System.Data;
using System.Reflection;

namespace RepTreeListHierarchicalDataRelationships {

    public class DataSetHelper {

        static private Type itemType;

        static public DataSet HierarchicalListToDataSet<T>(List<T> hlist, string keyFieldName, string parentFieldName) {
            DataSetHelper.itemType = typeof(T);
            DataSet dataSet = new DataSet();
            List<int> currentLevelIds = new List<int>(new int[] { 0 });
            List<int> nextLevelIds = new List<int>();
            int count = 0;

            do {
                DataTable dataTable = GetNewTable(DataSetHelper.itemType.Name + (count == 0 ? "RootTable" : "Table" + count.ToString()), keyFieldName);

                for(int i = 0; i < hlist.Count; i++) {
                    for(int j = 0; j < currentLevelIds.Count; j++) {

                        if(Convert.ToInt32(DataSetHelper.itemType.GetProperty(parentFieldName).GetValue(hlist[i], null)) == currentLevelIds[j]) {
                            AddItemToTable(dataTable, hlist[i]);
                            nextLevelIds.Add(Convert.ToInt32(DataSetHelper.itemType.GetProperty(keyFieldName).GetValue(hlist[i], null)));
                        }

                    }
                }

                if(nextLevelIds.Count > 0) {
                    dataSet.Tables.Add(dataTable);
                    if(count > 0) // Not a root table
                        dataSet.Relations.Add(string.Format("{0}Level{1}Relation", DataSetHelper.itemType.Name, count), dataSet.Tables[count - 1].Columns[keyFieldName], dataTable.Columns[parentFieldName]);
                }

                currentLevelIds.Clear();
                currentLevelIds.AddRange(nextLevelIds);
                nextLevelIds.Clear();
                count++;

            } while(currentLevelIds.Count > 0);

            return dataSet;
        }

        static private DataTable GetNewTable(string tableName, string keyFieldName) {
            DataTable dt = new DataTable(tableName);
            PropertyInfo[] properties = DataSetHelper.itemType.GetProperties();

            foreach(PropertyInfo property in properties) {
                dt.Columns.Add(property.Name, property.PropertyType);
            }

            dt.Constraints.Add("IDPK", dt.Columns[keyFieldName], true);

            return dt;
        }

        static private void AddItemToTable(DataTable dataTable, object item) {
            DataRow dataRow = dataTable.NewRow();
            PropertyInfo[] properties = DataSetHelper.itemType.GetProperties();

            foreach(PropertyInfo property in properties) {
                dataRow[property.Name] = property.GetValue(item, null);
            }

            dataTable.Rows.Add(dataRow);
        }

    }


    public class DataSetHelper2 {

        static public void FillTreeView(TreeView treeView, DataSet dataSet) {
            treeView.BeginUpdate();
            treeView.Nodes.Clear();

            for(int i = 0; i < dataSet.Tables.Count; i++) {
                treeView.Nodes.Add(dataSet.Tables[i].TableName);
                AddColumnsAndRelations(treeView.Nodes[i], dataSet.Tables[i]);
            }

            treeView.ExpandAll();
            treeView.EndUpdate();
        }

        static private void AddColumnsAndRelations(TreeNode dataTableTreeNode, DataTable dataTable) {
            DataSet dataSet = dataTable.DataSet;

            for(int i = 0; i < dataTable.Columns.Count; i++) {
                dataTableTreeNode.Nodes.Add(dataTable.Columns[i].ColumnName);
            }

            for(int i = 0; i < dataSet.Relations.Count; i++) {
                if(dataSet.Relations[i].ParentTable == dataTable) {
                    TreeNode relationNode = dataTableTreeNode.Nodes.Add(dataSet.Relations[i].RelationName);

                    AddColumnsAndRelations(relationNode, dataSet.Relations[i].ChildTable);
                }
            }
        }
        
    }

}

File: VB/HierarchyToDataSet.vb
Imports System
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports System.Data
Imports System.Reflection

Namespace RepTreeListHierarchicalDataRelationships

    Public Class DataSetHelper

        Shared Private itemType As Type

        Shared Public Function HierarchicalListToDataSet(Of T)(ByVal hlist As List(Of T), ByVal keyFieldName As String, ByVal parentFieldName As String) As DataSet
            itemType = GetType(T)
            Dim dataSet As DataSet = New DataSet()
            Dim currentLevelIds As List(Of Integer) = New List(Of Integer)(New Integer() {0})
            Dim nextLevelIds As List(Of Integer) = New List(Of Integer)()
            Dim count As Integer = 0
            Do
                Dim dataTable As DataTable = GetNewTable(itemType.Name & (If(count = 0, "RootTable", "Table" & count.ToString())), keyFieldName)
                For i As Integer = 0 To hlist.Count - 1
                    For j As Integer = 0 To currentLevelIds.Count - 1
                        If Convert.ToInt32(itemType.GetProperty(parentFieldName).GetValue(hlist(i), Nothing)) = currentLevelIds(j) Then
                            AddItemToTable(dataTable, hlist(i))
                            nextLevelIds.Add(Convert.ToInt32(itemType.GetProperty(keyFieldName).GetValue(hlist(i), Nothing)))
                        End If
                    Next
                Next

                If nextLevelIds.Count > 0 Then
                    dataSet.Tables.Add(dataTable)
                    If count > 0 Then dataSet.Relations.Add(String.Format("{0}Level{1}Relation", itemType.Name, count), dataSet.Tables(count - 1).Columns(keyFieldName), dataTable.Columns(parentFieldName)) ' Not a root table
                End If

                currentLevelIds.Clear()
                currentLevelIds.AddRange(nextLevelIds)
                nextLevelIds.Clear()
                count += 1
            Loop While currentLevelIds.Count > 0

            Return dataSet
        End Function

        Shared Private Function GetNewTable(ByVal tableName As String, ByVal keyFieldName As String) As DataTable
            Dim dt As DataTable = New DataTable(tableName)
            Dim properties As PropertyInfo() = itemType.GetProperties()
            For Each [property] As PropertyInfo In properties
                dt.Columns.Add([property].Name, [property].PropertyType)
            Next

            dt.Constraints.Add("IDPK", dt.Columns(keyFieldName), True)
            Return dt
        End Function

        Shared Private Sub AddItemToTable(ByVal dataTable As DataTable, ByVal item As Object)
            Dim dataRow As DataRow = dataTable.NewRow()
            Dim properties As PropertyInfo() = itemType.GetProperties()
            For Each [property] As PropertyInfo In properties
                dataRow([property].Name) = [property].GetValue(item, Nothing)
            Next

            dataTable.Rows.Add(dataRow)
        End Sub
    End Class

    Public Class DataSetHelper2

        Shared Public Sub FillTreeView(ByVal treeView As TreeView, ByVal dataSet As DataSet)
            treeView.BeginUpdate()
            treeView.Nodes.Clear()
            For i As Integer = 0 To dataSet.Tables.Count - 1
                treeView.Nodes.Add(dataSet.Tables(i).TableName)
                AddColumnsAndRelations(treeView.Nodes(i), dataSet.Tables(i))
            Next

            treeView.ExpandAll()
            treeView.EndUpdate()
        End Sub

        Shared Private Sub AddColumnsAndRelations(ByVal dataTableTreeNode As TreeNode, ByVal dataTable As DataTable)
            Dim dataSet As DataSet = dataTable.DataSet
            For i As Integer = 0 To dataTable.Columns.Count - 1
                dataTableTreeNode.Nodes.Add(dataTable.Columns(i).ColumnName)
            Next

            For i As Integer = 0 To dataSet.Relations.Count - 1
                If dataSet.Relations(i).ParentTable Is dataTable Then
                    Dim relationNode As TreeNode = dataTableTreeNode.Nodes.Add(dataSet.Relations(i).RelationName)
                    Call AddColumnsAndRelations(relationNode, dataSet.Relations(i).ChildTable)
                End If
            Next
        End Sub
    End Class
End Namespace
